### PR TITLE
Add build_only_index flag

### PR DESCRIPTION
**Pull request by maxulysse**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: lib/WorkflowSarek.groovy [hunk 1]
```diff
@@ -76,28 +76,30 @@ class WorkflowSarek {
...
-                                            break
-            default:    log.warn "Please provide an input samplesheet to the pipeline e.g. '--input samplesheet.csv'"
-                        exit 1, "Unknown step $params.step"
+        if (!params.build_only_index) {
+            switch (params.step) {
+                case 'mapping':                 log.warn "Can't start with step $params.step without samplesheet"
+                                                System.exit(1);
+                                                break
+                case 'markduplicates':          log.warn "Using file ${params.outdir}/csv/mapped.csv"
+                                                params.putIfAbsent("input","${params.outdir}/csv/mapped.csv");
```
Comment (matthdsm) [inline lib/WorkflowSarek.groovy:85]: `putIfAbsent`
Nice, didn't even know this existed!

File: nextflow_schema.json [hunk 2]
```diff
@@ -728,6 +728,12 @@
                     "description": "Save built references.",
                     "help_text": "Set this parameter, if you wish to save all computed reference files. This is useful to avoid re-computation on future runs."
                 },
+                "build_only_index": {
+                    "type": "boolean",
+                    "fa_icon": "fas fa-download",
+                    "description": "Only built references.",
+                    "help_text": "Set this parameter, if you wish to comput and save all computed reference files. This is useful to avoid re-computation on future runs."
```
Comment (matthdsm) [inline nextflow_schema.json:735]:
```suggestion
                    "help_text": "Set this parameter, if you wish to compute and save all computed reference files. This is useful to avoid re-computation on future runs."
```

File: subworkflows/local/prepare_genome/main.nf [hunk 3]
```diff
@@ -27,31 +27,32 @@ include { UNZIP as UNZIP_RT                      } from '../../../modules/nf-cor
...
 
 
     main:
+    fasta = fasta.map{ fasta -> [ [ id:fasta.baseName ], fasta ] }
+    versions = Channel.empty()
 
-    ch_versions = Channel.empty()
+    BWAMEM1_INDEX(fasta)     // If aligner is bwa-mem
+    BWAMEM2_INDEX(fasta)     // If aligner is bwa-mem2
+    DRAGMAP_HASHTABLE(fasta) // If aligner is dragmap
```
Comment (matthdsm) [inline subworkflows/local/prepare_genome/main.nf:51]: Still no love for `snap-aligner` 🥺 ?

File: subworkflows/local/prepare_genome/main.nf [hunk 4]
```diff
@@ -27,31 +27,32 @@ include { UNZIP as UNZIP_RT                      } from '../../../modules/nf-cor
...
+    fasta_fai         // channel: [optional]  fasta_fai
+    germline_resource // channel: [optional]  germline_resource
+    known_indels      // channel: [optional]  known_indels
+    known_snps        // channel: [optional]  known_snps
+    pon               // channel: [optional]  pon
 
 
     main:
+    fasta = fasta.map{ fasta -> [ [ id:fasta.baseName ], fasta ] }
+    versions = Channel.empty()
```
Comment (matthdsm) [inline subworkflows/local/prepare_genome/main.nf:46]: Why the change to `versions`?

**Review by matthdsm — APPROVED** [review]
Some small comments and questions, but LGTM!

File: nextflow_schema.json (hunk 2 above)
Comment (jfy133) [reply nextflow_schema.json:735]: >  if you wish to compute and save all computed reference files

Don't you mean 'if you wish to only compute and save all computed reference files. No mapping or downstream steps will be performed.

File: subworkflows/local/prepare_genome/main.nf (hunk 4 above)
Comment (jfy133) [reply subworkflows/local/prepare_genome/main.nf:46]: Agreed? I don't think that's the (implied) 'standard'?

**Review by jfy133 — APPROVED** [review]

File: subworkflows/local/prepare_genome/main.nf (hunk 3 above)
Comment (maxulysse) [reply subworkflows/local/prepare_genome/main.nf:51]: We need more time for that, but it'll come.
Do we have a github issue for that?

File: subworkflows/local/prepare_genome/main.nf (hunk 4 above)
Comment (maxulysse) [reply subworkflows/local/prepare_genome/main.nf:46]: I feel like most of the stuff we have in DSL2 are channels, I think it makes more sense to prefix everything else.

File: subworkflows/local/prepare_genome/main.nf (hunk 3 above)
Comment (matthdsm) [reply subworkflows/local/prepare_genome/main.nf:51]: Don't think so, not all that important too ;)

File: subworkflows/local/prepare_genome/main.nf (hunk 4 above)
Comment (matthdsm) [reply subworkflows/local/prepare_genome/main.nf:46]: I personally like knowing at a glance what's a channel and what's a value, but that's me 🤷🏻 The other way around accomplished the same thing.

Comment (jfy133) [reply subworkflows/local/prepare_genome/main.nf:46]: I'm wiht @matthdsm here. I quite like the SBWF guidelines on this: https://nf-co.re/docs/contributing/subworkflows#naming-conventions

**Review by nvnieuwk — APPROVED** [review]

**Comment by nvnieuwk** [comment]
Again the conda tests though 🤔

**Comment by maxulysse** [comment]
> Again the conda tests though thinking

conda is super flaky on GHA lately

**Comment by nvnieuwk** [comment]
Yes it's weird...

File: subworkflows/local/prepare_genome/main.nf (hunk 4 above)
Comment (maxulysse) [reply subworkflows/local/prepare_genome/main.nf:46]: I'll have a look at it for the next sarek refactoring, and I'll ping you both for reviews

File: subworkflows/local/prepare_genome/main.nf (hunk 3 above)
Comment (FriederikeHanssen) [reply subworkflows/local/prepare_genome/main.nf:51]: sorry no, more important at the moment to make sure the existing aligners all play nice with the variantcallers we have. If you open an issue we can pencil it in eventually
